develop/blockchains/custom-blockchains/benchmarking.md

```diff
@@ -1,4 +1,191 @@
+
+Within FRAME, each function call that is dispatched must have a `#[weight]` annotation that can return the expected weight for the worst-case scenario execution of that function given its inputs. The benchmarking framework automatically generates a file with those formulas for you, which you can use in that annotation within your pallet.
+
+## Benchmarking Steps
+
+Benchmarking a pallet involves the following steps: 
+
+1. Creating a `benchmarking.rs` file within your pallet's structure
+2. Writing a benchmarking test for each extrinisic
+3. Executing the benchmarking tool, producing a linear model that measures the weight (time and space for an extrinsic)
+
+The benchmarking tool then runs multiple iterations with different possible parameters for each of these functions, wherein the approximate **worst** case scenario is run and applied as the weight. In usage, if the weight is better than this benchmark, fees are adjusted and refunded accordingly. By default, the benchmarking pipeline is deactivated, and it can be enabled by passing in the `runtime-benchmarks` feature flag when compiling your runtime.
+
+## Preparing Your Environment
+
+Before writing benchmarks, you need to ensure the `frame-benchmarking` crate is included in your pallet's `Cargo.toml`:
+
+```toml
+frame-benchmarking = { version = "37.0.0", default-features = false }
+```
+
+You must also ensure that you add the `runtime-benchmarks` feature flag as follows under the `[features]` section of your pallet's `Cargo.toml`:
+
+```toml
+runtime-benchmarks = [
+    "frame-benchmarking/runtime-benchmarks",
+    "frame-support/runtime-benchmarks",
+    "frame-system/runtime-benchmarks",
+    "sp-runtime/runtime-benchmarks",
+]
+```
+
+Lastly, ensure that `frame-benchmarking` is included in `std = []`: 
+
+```toml
+std = [
+    # ...
+    "frame-benchmarking?/std",
+    # ...
+]
+```
+
+Once this is complete, you have the required dependencies for writing benchmarks for your pallet.
+
+## Writing Benchmarks
+
+Create a `benchmarking.rs` file in your pallet's `src/`. You may copy the barebones template below to get started: 
+
+!!!note "This example is from the pallet template"
+ Take a look at the repository in the [`polkadot-sdk-parachain-template`](https://github.com/paritytech/polkadot-sdk-parachain-template/tree/master/pallets){target=_blank}
+
+```rust
+//! Benchmarking setup for pallet-template
+#![cfg(feature = "runtime-benchmarks")]
+
+use super::*;
+use frame_benchmarking::v2::*;
+
+#[benchmarks]
+mod benchmarks {
+    use super::*;
+ #[cfg(test)]
+    use crate::pallet::Pallet as Template;
+    use frame_system::RawOrigin;
+
+ #[benchmark]
+    fn do_something() {
+        let caller: T::AccountId = whitelisted_caller();
+ #[extrinsic_call]
+ do_something(RawOrigin::Signed(caller), 100);
+ assert_eq!(Something::<T>::get().map(|v| v.block_number), Some(100u32.into()));
+ }
+
+ impl_benchmark_test_suite!(Template, crate::mock::new_test_ext(), crate::mock::Test);
+}
```

Comment:
This should be a code snippet, or you can pull the source code from the actual parachain template
```suggestion
```rust
--8<-- 'https://raw.githubusercontent.com/paritytech/polkadot-sdk-parachain-template/refs/tags/v0.0.2/pallets/template/src/benchmarking.rs'
```
```

Reply:
kind of a crazy-ish idea - what if we just added basically the entire solochain / parachain template code as part of snippets? That way we just grab what we need from it, also helps in the future if we need to compile stuff

Reply:
I like that idea @CrackTheCode016 . Even for those sections for which we only grab portions of the code, we can split it with the plugin we are using, for example:
```rust
--8<-- 'snippets/solochain-template.rs:10:14'
/ we can add comments here if we need
--8<-- 'snippets/solochain-template.rs:17:20'
```
 However, I'm just wondering about the following:
 1. Would having an entire Rust project in the snippets folder affect the rendering of the page somehow? Maybe @eshaben can provide some input here.
 2. What if we need different versions of the solochain/parachain template to showcase different things? Would we have different projects? Like: solochain-for-testing.rs, solochain-for-benchamarking.rs, and so on..

Reply:
I would say we could have one "kitchensink" runtime / template that has everything (integration tests, unit tests, benchmarking, any custom pallet stuff, signed extensions.. etc..) which we use for this purpose. Yeah, we may need a solo and para template respectively, dunno for sure yet

This way, we can also address the matter of having one compilable repository as @kianenigma mentioned.  Another approach is to separate them I guess, but I like the idea of a cohesive example

Reply:
FYI, let's do this in another PR / after some discussion, but if we do it we can update all of these FRAME examples to use that instead.

Then, we can have a CI job which compiles / runs it every so often, ensuring it is up to date and works, then if it does break we can easily update it as needed and the examples should be mostly intact. How does that sound?